quantity/core.py

```diff
@@ -0,0 +1,255 @@
+
+    def deferred(self):
+        return attr_getter(self.value)
+
+    return property(deferred)
+
+
+def _make_same_unit_method(attr):
+    if array_api_func := getattr(array_api_compat, attr, None):
+
+        def same_unit(self, *args, **kwargs):
+            return replace(self, value=array_api_func(self.value, *args, **kwargs))
+
+    else:
+
+        def same_unit(self, *args, **kwargs):
+            return replace(self, value=getattr(self.value, attr)(*args, **kwargs))
+
+    return same_unit
+
+
+def _make_same_unit_attribute(attr):
+    attr_getter = getattr(array_api_compat, attr, operator.attrgetter(attr))
+
+    def same_unit(self):
+        return replace(self, value=attr_getter(self.value))
+
+    return property(same_unit)
+
+
+def _make_defer_dimensionless(attr):
+    def defer_dimensionless(self):
+        try:
+            return getattr(self.unit.to(DIMENSIONLESS, self.value), attr)()
+        except Exception as exc:
+            raise TypeError from exc
+
+    return defer_dimensionless
+
+
+def _check_pow_args(exp, mod):
+    if mod is not None:
+        return NotImplemented
+
+    if not isinstance(exp, PYTHON_NUMBER):
+        try:
+            exp = exp.__complex__()
+        except Exception:
+            try:
+                return exp.__float__()
+            except Exception:
+                return NotImplemented
+
+    return exp.real if exp.imag == 0 else exp
+
+
+@dataclass(frozen=True, eq=False)
+class Quantity:
+    value: Any
+    unit: u.UnitBase
+
+    def __array_namespace__(self, *, api_version: str | None = None) -> Any:
+        # TODO: make our own?
+        return np
+
+    def _operate(self, other, op, units_helper):
+        if not has_array_namespace(other) and not isinstance(other, PYTHON_NUMBER):
+            # HACK: unit should take care of this!
+            if not isinstance(other, u.UnitBase):
+                return NotImplemented
+
+            try:
+                unit = getattr(self.unit, op.replace("__i", "__"))(other)
+            except Exception:
+                return NotImplemented
+            else:
+                return replace(self, unit=unit)
+
+        other_value, other_unit = get_value_and_unit(other)
+        self_value = self.value
+        (conv0, conv1), unit = units_helper(self.unit, other_unit)
+        if conv0 is not None:
+            self_value = conv0(self_value)
+        if conv1 is not None:
+            other_value = conv1(other_value)
+        try:
+            value = getattr(self_value, op)(other_value)
+        except AttributeError:
+            return NotImplemented
+        if value is NotImplemented:
+            return NotImplemented
+
+        return replace(self, value=value, unit=unit)
+
+    # Operators (skipping ones that make no sense, like __and__);
+    # __pow__ and __rpow__ need special treatment and are defined below.
+    __add__, __radd__, __iadd__ = _make2arg_ops("__add__")
```

Comment:
As followup, we should get `mypy` set up to do static testing to make sure it's happy with this magic.